src/Domain/Storage/ImageStore.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Liip\ImagineBundle\Domain\Storage;
+
+use Liip\ImagineBundle\Domain\ImageReference;
+
+interface ImageStore
+{
+    public function exists(string $imageId, string $stackName, string $targetFormat): bool;
+    public function getUrl(string $imageId, string $stackName, string $targetFormat): string;
+    public function store(ImageReference $image, string $imageId, string $stackName, string $targetFormat): ImageReference;
+
+    /**
+     * Delete a cached transformed image.
+     *
+     * Always deletes the image in all formats in which it exists.
+     *
+     * @param string[] $stackNames Remove cached images with specified stack names. If empty, all stacks are deleted.
+     */
+    public function delete(string $imageId, array $stackNames = []): void;
+
+    /**
+     * Whether this store can create the image on the fly when the URL it returns is fetched.
+     *
+     * (e.g. the URL points to a controller or a 404 handler is used to generate the image on demand)
+     */
+    public function supportsOnDemandCreation(): bool;
```

Comment:
i added this to handle both scenarios of having to generate the image when requesting the url or capable of generating on demand.

we could do a separate interface for it (but would need the method anyways because depending on the store this option could depend on whether configuration is set up)